### PR TITLE
ARROW-5350: [Rust] Allow filtering on simple lists

### PR DESCRIPTION
**Pull request by nevi-me**
This extends filters to simple lists.  CC @yordan-pavlov

### CODE REVIEW

File: rust/arrow/src/compute/kernels/filter.rs [hunk 1]
```diff
@@ -230,6 +231,86 @@ macro_rules! filter_dictionary_array {
...
+macro_rules! filter_primitive_item_list_array {
+    ($context:expr, $array:expr, $item_type:ident) => {{
+        let input_array = $array.as_any().downcast_ref::<ListArray>().unwrap();
+        let values_builder = PrimitiveBuilder::<$item_type>::new($context.filtered_count);
+        let mut builder = ListBuilder::new(values_builder);
+        for i in 0..$context.filter_u64.len() {
+            // foreach u64 batch
+            let filter_batch = $context.filter_u64[i];
+            if filter_batch == 0 {
+                // if batch == 0: skip
```
Comment (alamb) [inline rust/arrow/src/compute/kernels/filter.rs:243]: I think this comment would be improved if it explained "why" batch zero is being skipped (the code is pretty clear that batch 0 *is* being skipped). Same comment for `filter_non_primitive_item_list_array`

File: rust/arrow/src/compute/kernels/filter.rs [hunk 2]
```diff
@@ -230,6 +231,86 @@ macro_rules! filter_dictionary_array {
...
+                    }
+                }
+            }
+        }
+        Ok(Arc::new(builder.finish()))
+    }};
+}
+
+macro_rules! filter_non_primitive_item_list_array {
+    ($context:expr, $array:expr, $item_array_type:ident, $item_builder:ident) => {{
```
Comment (alamb) [inline rust/arrow/src/compute/kernels/filter.rs:275]: I like this "item_builder" pattern for non-primative list arrays

File: rust/arrow/src/compute/kernels/filter.rs [hunk 3]
```diff
@@ -733,4 +909,53 @@ mod tests {
...
+        assert_eq!(DataType::Int32, d.value_type());
+
+        // result should be [[3, 4, 5], null]
+        assert_eq!(2, d.len());
+        assert_eq!(1, d.null_count());
+        assert_eq!(true, d.is_null(1));
+
+        assert_eq!(0, d.value_offset(0));
+        assert_eq!(3, d.value_length(0));
+        assert_eq!(3, d.value_offset(1));
```
Comment (alamb) [inline rust/arrow/src/compute/kernels/filter.rs:945]: I wonder if it is necessary to check `value_offset` and `value_length` for an element in the list that is marked as null?

File: rust/arrow/src/compute/kernels/filter.rs [hunk 4]
```diff
@@ -733,4 +909,53 @@ mod tests {
...
+        assert_eq!(0, d.value_length(1));
+        assert_eq!(
+            Buffer::from(&[3, 4, 5].to_byte_slice()),
+            d.values().data().buffers()[0].clone()
+        );
+        assert_eq!(
```
Comment (alamb) [inline rust/arrow/src/compute/kernels/filter.rs:951]: I am not sure what the significance of checking the data buffer is here. It seems like the contents are covered in the other checks.

**Review by alamb — COMMENTED** [review]
Looks good to me -- I think it is worth a test for non-primitive arrays (e.g. Utf8) but otherwise I think this PR is 👍 👍  . I left some other minor comments.

**Review by jorgecarleitao — COMMENTED** [review]
I have the same comment as @alamb wrt to the why the 0 is being skipped, but other wise LGTM!

File: rust/arrow/src/compute/kernels/filter.rs (hunk 1 above)
Comment (yordan-pavlov) [reply rust/arrow/src/compute/kernels/filter.rs:243]: the meaning of this condition is not that batch 0 is skipped, but rather that if the current batch (of 64 bits / elements from the filter array) contains only 0s (i.e. all items are filtered out) then the entire batch should be skipped as there is no items to copy to the output array (all items are filtered out); this continues until a batch is encountered where at least one item is not filtered out (and therefore the numeric representation of the batch of filter values is != 0); this skipping of entire batches of 64 items is a major factor to the performance improvements recently implemented in the filter kernel

Comment (alamb) [reply rust/arrow/src/compute/kernels/filter.rs:243]: This makes total sense @yordan-pavlov  -- I was suggesting an improvement to the comment itself. Perhaps something like:

```suggestion
                // if batch == 0, all items are filtered out, so skip entire batch
```

**Review by alamb — APPROVED** [review]

File: rust/arrow/src/compute/kernels/filter.rs [hunk 5]
```diff
@@ -399,6 +480,99 @@ impl FilterContext {
...
+                DataType::Timestamp(TimeUnit::Nanosecond, _) => {
+                    filter_primitive_item_list_array!(self, array, TimestampNanosecondType)
+                }
+                DataType::Binary => filter_non_primitive_item_list_array!(
+                    self,
+                    array,
+                    BinaryArray,
+                    BinaryBuilder
+                ),
+                DataType::Utf8 => filter_non_primitive_item_list_array!(
```
[the remark this answers is not in the document]
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/filter.rs:563]: I've added largeutf8, largebinary and largelist. I'll work on test coverage separately

File: rust/arrow/src/compute/kernels/filter.rs (hunk 4 above)
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/filter.rs:951]: It's a way of checking the values. From what I see, the original author of the PR (not me lol) checked the lengths, then the values by using the buffers.
I think it's fine.

File: rust/arrow/src/compute/kernels/filter.rs (hunk 3 above)
Comment (nevi-me) [reply rust/arrow/src/compute/kernels/filter.rs:945]: I suppose it's a function of how much we can accept as working vs what additional functionality we want to test; e.g. we've tested in list arrays that null offsets behave correctly.
My preference would be to forego some of the assertions that for example a value that should be null is null, if the functionality isn't expected to change such behaviour.

**Review by jorgecarleitao — APPROVED** [review]
LGTM. Thanks a lot for taking the time to port this from another PR, @nevi-me  (pending CI)

File: rust/arrow/src/compute/kernels/filter.rs (hunk 4 above)
Comment (alamb) [reply rust/arrow/src/compute/kernels/filter.rs:1088]: I think it is reasonable to keep the tests consistent

**Pull request closed without merging**